### PR TITLE
Fix Error when Authed, but deleted

### PR DESCRIPTION
**Pull request by ryuyu**
In the case when a session was authenticated as a user, but the user was deleted BEFORE the session was logged out, every subsequent to the gallery would 500.

This fix makes it so if we get into this state, the user will be logged out, redirected to the home page, and shown a message indicating that their account was deleted.

Preview:
image

### CODE REVIEW

File: src/NuGetGallery/Strings.resx [hunk 1]
```diff
@@ -1141,4 +1141,7 @@ The {1} Team</value>
   <data name="UploadPackage_UnsupportedIconImageFormat" xml:space="preserve">
     <value>Unsupported icon image format. Only PNG and JPG images are supported.</value>
   </data>
+  <data name="LoggedInUserDeleted" xml:space="preserve">
+    <value>Your account is no longer active and we have logged you out. Please contact support if this is unexpected.</value>
```
Comment (joelverhagen) [inline src/NuGetGallery/Strings.resx]: Should we put the support email in there? i think there is a config value for it.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public sealed class UserDeletedErrorFilter : FilterAttribute, IExceptionFilter
+    {
+        public void OnException(ExceptionContext context)
+        {
+            if (context.Exception is UserDeletedException)
+            {
+                context.HttpContext.Response.Clear();
+                context.HttpContext.Response.TrySkipIisCustomErrors = true;
+                context.HttpContext.Response.StatusCode = 401;
```
Comment (joelverhagen) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: why set 401? I guess the `RedirectResult` below is taking precedence.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+                context.Controller.TempData = new TempDataDictionary
+                {
+                    { "Message", Strings.LoggedInUserDeleted }
+                };
+
+                context.Result = new RedirectResult("/");
```
Comment (joelverhagen) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: This is a temporary redirect right? :)

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+using System.Web.Mvc;
+
+namespace NuGetGallery.Infrastructure
+{
+    public sealed class UserDeletedErrorFilter : FilterAttribute, IExceptionFilter
+    {
+        public void OnException(ExceptionContext context)
+        {
+            if (context.Exception is UserDeletedException)
```
Comment (joelverhagen) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: Can we verify that the user is indeed logged in at the beginning of the request?

File: src/NuGetGallery/Strings.resx (hunk 1 above)
Comment (scottbommarito) [inline src/NuGetGallery/Strings.resx]: I think `no longer active` is not clear enough. People are going to think that we have different concepts of "inactive" and "deleted" and that's simply not true. If somebody's account was unintentionally deleted, they need to know that it was so they can reach out to us as soon as possible, and using the term "inactive" might confuse people. I think this should just say `Your account was deleted`.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs (hunk 4 above)
Comment (scottbommarito) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: I think this case is exceptional enough that we should have some telemetry about it.  I know we have another error filter that logs exceptions. We should have at least some sort of log here.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs (hunk 3 above)
Comment (scottbommarito) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: What's the default behavior when we return 401? Doesn't it redirect to sign in? Should we consider doing that instead of redirecting to the home page?

File: src/NuGetGallery.Services/AccountManagement/UserDeletedException.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+using System.Web;
+
+namespace NuGetGallery
+{
+    public class UserDeletedException : HttpException
```
Comment (scottbommarito) [inline src/NuGetGallery.Services/AccountManagement/UserDeletedException.cs]: I think we can have a more clear name here.

`CurrentUserWasDeleted` or `LoggedInUserWasDeleted` are clearer about the problem. Don't forget to change the name of the error filter as well.

Also, consider adding some XML docs to this class.

File: src/NuGetGallery.Services/Extensions/IOwinContextExtensions.cs [hunk 6]
```diff
@@ -47,7 +47,8 @@ public static User GetCurrentUser(this IOwinContext self)
             {
                 // Unauthorized! If we get here it's because a valid session token was presented, but the
                 // user doesn't exist any more. So we just have a generic error.
-                throw new HttpException(401, ServicesStrings.Unauthorized);
+                self.Authentication.SignOut();
```
Comment (scottbommarito) [inline src/NuGetGallery.Services/Extensions/IOwinContextExtensions.cs]: Could you move this sign out call to the error filter?

The reason I say this is that regardless of where we discover the current user was deleted, we should always log them out. If we discover other code paths in which we find the current user has been deleted, we shouldn't have to add this sign out call there too.

Not a big deal though, as it's unlikely we'll find another place.

**Comment by skofman1** [comment]
How was this tested?

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs (hunk 2 above)
Comment (ryuyu) [reply src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: It probably is. This was a vestige of something i was trying earlier. It should be safe to remove it entirely.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs (hunk 3 above)
Comment (ryuyu) [reply src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: Looks like it redirects with 302, so should be temporary.

Comment (ryuyu) [reply src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: Removing the 401 status set here.

File: src/NuGetGallery.Services/Extensions/IOwinContextExtensions.cs (hunk 6 above)
Comment (ryuyu) [reply src/NuGetGallery.Services/Extensions/IOwinContextExtensions.cs]: Probably not. We don't have access to the IOwinContext inside the error filter. I imagine we could emulate it by manually clearing cookies, but I find making this call (the one we use other places to log out) more comfortable.

Comment (scottbommarito) [reply src/NuGetGallery.Services/Extensions/IOwinContextExtensions.cs]: Agreed. Was just curious if you could.

File: src/NuGetGallery.Services/AccountManagement/UserDeletedException.cs (hunk 5 above)
Comment (ryuyu) [reply src/NuGetGallery.Services/AccountManagement/UserDeletedException.cs]: UserDeletedException -> CurrentUserDeletedException

File: src/NuGetGallery/Strings.resx (hunk 1 above)
Comment (ryuyu) [reply src/NuGetGallery/Strings.resx]: Technically, if a user logged in with MSA/AAD and the account was deleted, they can still log in with the same MSA/AAD and it will link to a new NuGet Account. I think "no longer active" better describes this state (ie the specific NuGet Account that was part of your identity, isn't there anymore for one reason or another), but I don't feel strongly about it. I will defer to @anangaur @karann-msft

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs (hunk 4 above)
Comment (ryuyu) [reply src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: They won't be at this point. We log them out before we throw the exception.
The GetCurrentUser extension method on the IOwinContext already verifies that a identity was present to us at the beginning of the call.

TL;DR:
If there was no logged in session, we would never hit this path to begin with.

File: src/NuGetGallery/Strings.resx (hunk 1 above)
Comment (scottbommarito) [reply src/NuGetGallery/Strings.resx]: That's an interesting edge-case, but I still believe `Your NuGet account was deleted` is the best thing to say there. It removes any confusion over MSA vs AAD vs NuGet while also communicating what happened.

**Comment by ryuyu** [comment]
Tested locally for both password auth and MSA auth accounts.

**Comment by scottbommarito** [comment]
Does this PR fix https://github.com/NuGet/NuGetGallery/issues/7033?

File: src/NuGetGallery/Strings.resx (hunk 1 above)
Comment (loic-sharma) [reply src/NuGetGallery/Strings.resx]: Also, consider inverting the first sentence: `You have been logged out as ...`

Comment (ryuyu) [reply src/NuGetGallery/Strings.resx]: updated.

Comment (shishirx34) [reply src/NuGetGallery/Strings.resx]: I don't suppose we have the state of the user deletion? as in how was the account deleted? was it DSR or user request to delete. If we can specify the reason it will be helpful in customer requests if someone were to see this and dispute.

Comment (ryuyu) [reply src/NuGetGallery/Strings.resx]: We have no idea at this point. We just know that we got a valid user session, and the user it specified didn't exist.

File: src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+using System.Web.Mvc;
+
+namespace NuGetGallery.Infrastructure
+{
+    public sealed class UserDeletedErrorFilter : FilterAttribute, IExceptionFilter
```
Comment (scottbommarito) [inline src/NuGetGallery/Infrastructure/UserDeletedErrorFilter.cs]: Rename this to `CurrentUserDeletedErrorFilter`?
